notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
-    # set _scheme='https' explicitly.
-    protocol = request.headers.get('X-Forwarded-Proto', None)
-    url = flask.url_for('root', _scheme=protocol, _external='true')
-    return url + path
+def del_pod(pod_name: str):
+    k8s.delete_namespaced_pod(
+        pod_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+def del_svc(svc_name: str):
+    k8s.delete_namespaced_service(
+        svc_name,
+        'default',
+        kube.client.V1DeleteOptions(),
+        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+
+
+# General resource marshalling functions #
+
+
+def get_path(data, path: str):
+    # Not using tail recursion because python doesn't optimize such calls
+    while True:
+        idx = path.find('.')
+
+        if idx == -1:
+            if isinstance(data, dict):
+                return data[path]
+            return getattr(data, path)
+
+        data = getattr(data, path[0:idx])
 
+        path = path[idx + 1:]
+
+
+def marshall_json(resources: [], paths=[], flatten=False):
+    if len(resources) == 0 or len(paths) == 0:
+        if flatten:
+            return "{}"
+
+        return "[]"
+
+    resp = []
+    for rsc in resources:
+        data = {}
+        for path in paths:
+            if len(path) == 3:
+                data[path[0]] = path[2](get_path(rsc, path[1]))
+            else:
+                data[path[0]] = get_path(rsc, path[1])
+
+        resp.append(data)
+
+    if flatten and len(resources) == 1:
+        return ujson.dumps(resp[0])
+
+    return ujson.dumps(resp)
+
+
+def read_svc_status(svc_name: str):
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
+    """
+        Summarize the container status based on its most recent state
+
+        Parameters
+        ----------
+        container_statuses : list[V1ContainerStatus]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md
+    """
+    if container_statuses is None:
+        return None
+
+    state = container_statuses[0].state
+    rn = None
+    wt = None
+    tm = None
+    if state.running:
+        rn = {"started_at": state.running.started_at}
+
+    if state.waiting:
+        wt = {"reason": state.waiting.reason}
+
+    if state.terminated:
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
+    """
+        Return the most recent status=="True" V1PodCondition or None
```

Comment:
Our documentation style does not include an extra four spaces of indentation.

Reply:
What is your documentation style? Do you have a link?